src/core/connections.jl

```diff
@@ -0,0 +1,423 @@
+"""
+function connected_params(md::ModelDef, comp_name::Symbol)
+    ext_connections = Iterators.filter(x -> x.comp_name == comp_name, external_param_conns(md))
+    ext_set_params = map(x->x.param_name, ext_connections)
+
+    int_connections = Iterators.filter(x -> x.dst_comp_name == comp_name, internal_param_conns(md))
+    int_set_params = map(x->x.dst_par_name, int_connections)
+
+    return union(ext_set_params, int_set_params)
+end
+
+"""
+    unconnected_params(md::ModelDef)
+
+Return a list of tuples (componentname, parametername) of parameters
+that have not been connected to a value in the model.
+"""
+function unconnected_params(md::ModelDef)
+    unconnected = Vector{Tuple{Symbol,Symbol}}()
+    
+    for comp_def in compdefs(md)
+        comp_name = name(comp_def)
+        params = parameter_names(comp_def)
+        connected = connected_params(md, comp_name)
+        append!(unconnected, map(x->(comp_name, x), setdiff(params, connected)))
+    end
+
+    return unconnected
+end
+
+"""
+    set_leftover_params(m::Model, parameters::Dict{Any,Any})
+
+Set all the parameters in a model that don't have a value and are not connected
+to some other component to a value from a dictionary. This method assumes the dictionary
+keys are strings that match the names of unset parameters in the model.
+"""
+function set_leftover_params(md::ModelDef, parameters::Dict{String,Any})
+    parameters = Dict(lowercase(k) => v for (k, v) in parameters)
+    leftovers = unconnected_params(md)
+    external_params = md.external_params
+
+    for (comp_name, param_name) in leftovers
+        # check whether we need to set the external parameter
+        if ! haskey(md.external_params, param_name)
+            value = parameters[lowercase(string(param_name))]
+            param_dims = parameter_dimensions(md, comp_name, param_name)
+            num_dims = length(param_dims)
+
+            if num_dims == 0    # scalar case
+                set_external_scalar_param(md, param_name, value)
+
+            else
+                if num_dims in (1, 2) && param_dims[1] == :time   # array case
+                    value = convert(Array{md.numberType}, value)
+                    # start = indexvalues(md, :time)[1]
+                    start = dim_values(md, :time)[1]
+                    step = step_size(md)
+                    T = eltype(value)
+                    values = get_timestep_instance(T, start, step, num_dims, value)
+                else
+                    values = value
+                end
+                set_external_array_param(md, param_name, values, param_dims)
+            end
+        end
+        connect_parameter(md, comp_name, param_name, param_name)
+    end
+    nothing
+end
+
+internal_param_conns(md::ModelDef) = md.internal_param_conns
+
+external_param_conns(md::ModelDef) = md.external_param_conns
+
+function external_param(md::ModelDef, name::Symbol)
+    try
+        return md.external_params[name]
+    catch
```

Comment:
In general, I agree. But what else could raise an error in this case? The alternative is less efficient: test name in keys, then access it. Since we expect it to be there most of the time, why not just try it?

Reply:
Changed to
```
    try
        return md.external_params[name]
    catch err
        if err is KeyError
            error("$name not found in external parameter list")
        else
            rethrow(err)
    end
```